### PR TITLE
Salesforce Real-Time Monitoring Event Types EventStore Support

### PR DESCRIPTION
**Pull request by brianbeklaviyo**
# Salesforce Threat Detection Connector Enhancement

## Overview
This PR adds a comprehensive Salesforce Threat Detection connector that enables collection of [Shield Event Monitoring data](https://developer.salesforce.com/docs/atlas.en-us.256.0.platform_events.meta/platform_events/platform_events_objects_monitoring.htm). The connector significantly enhances Grove's Salesforce integration with comprehensive support for all Salesforce Shield EventStore types, robust error handling, and improved authentication methods.

### 🔒 **Comprehensive EventStore Support**
- **8 EventStore Types**: ApiAnomaly, SessionHijacking, CredentialStuffing, ReportAnomaly, BulkApiResult, GuestUserAnomaly, LoginAnomaly, PermissionSet
- **Dynamic Field Inclusion**: Uses `SELECT FIELDS(ALL)` to automatically capture all available fields
- **SOQL Compliance**: Added `LIMIT 200` to comply with Salesforce `FIELDS(ALL)` requirements

### 🛡️ **Enhanced Error Handling**
- **INVALID_TYPE Errors**: Detailed troubleshooting guidance for Shield licensing/permissions issues
- **INVALID_FIELD Errors**: Clear instructions for field-level permission problems
- **Shield Availability Check**: Proactive validation before data collection
- **Comprehensive Error Messages**: Step-by-step resolution guidance for administrators

### 🔐 **Authentication Improvements**
- **OAuth 2.0 Client Credentials**: Full support with proper URL selection
- **Legacy Username/Password**: Maintained backward compatibility
- **Mixed Configuration**: OAuth takes precedence when both methods are configured
- **Enhanced Logging**: Secure credential logging with proper redaction

### ⚡ **Performance & Reliability**
- **Rate Limiting**: Exponential backoff with jitter for retry logic
- **Robust Retry Logic**: Handles `REQUEST_LIMIT_EXCEEDED` and `QUERY_TIMEOUT` errors
- **Broader Time Range Fallback**: Automatically tries 30-day range if no recent data found
- **Pointer Management**: Improved timestamp handling with tie-breaking by ID

### 🧪 **Comprehensive Testing**
- **22 Test Cases**: Complete coverage of all functionality
- **PII Sanitization**: All test data uses obviously fake credentials
- **Mock Responses**: Proper API mocking for all authentication methods
- **Error Scenarios**: Tests for all error conditions and edge cases

### 📊 **Data Processing**
- **Dynamic Field Mapping**: Automatically includes all fields from EventStore records
- **Timestamp Parsing**: Handles both Z and timezone offset formats
- **Chronological Ordering**: Maintains proper event ordering
- **Metadata Enrichment**: Adds Grove-specific metadata to all entries

### 🔧 **Code Quality**
- **Type Hints**: Full type annotation coverage
- **Error Handling**: Comprehensive exception handling with specific error types
- **Logging**: Detailed debug and error logging throughout
- **Documentation**: Extensive docstrings and inline comments

### Core Connector
- `grove/connectors/sf/threat_detection.py` - Enhanced connector implementation

### Test Suite
- `tests/test_connectors_sf_threat_detection.py` - Comprehensive test coverage
- `tests/fixtures/sf/threat_detection/api_anomaly_direct.json` - ApiAnomaly test data
- `tests/fixtures/sf/threat_detection/report_anomaly_direct.json` - ReportAnomaly test data

## Breaking Changes
None - This is a backward-compatible enhancement.

## Dependencies
- Requires Salesforce Shield license for full functionality
- Uses existing Grove dependencies (no new requirements)

## Testing
- ✅ All 22 tests passing
- ✅ PII sanitization verified
- ✅ Linting checks passed
- ✅ Type checking passed

## Usage Example

```json
{
  "client_id": "YOUR_CONNECTED_APP_CLIENT_ID",
  "client_secret": "YOUR_CONNECTED_APP_CLIENT_SECRET",
  "identity": "USERNAME_HERE",
  "instance_url": "https://your-instance.my.salesforce.com",
  "name": "salesforce-threat-detection",
  "connector": "sf_threat_detection",
  "operation": "ApiAnomaly",
  "start_date": "2024-01-01T00:00:00.000Z"
}
```

## Error Handling Examples

The connector now provides detailed error messages for common issues:

- **Shield Not Available**: Clear instructions for enabling Event Monitoring
- **Field Permissions**: Guidance for checking field-level access
- **Authentication Issues**: Specific error messages for OAuth vs legacy auth problems

## Security Considerations
- All test data uses obviously fake credentials
- No PII or sensitive data in code or tests
- Proper credential redaction in logs
- Secure OAuth implementation following best practices

### CODE REVIEW

**Comment by hcpadkins** [comment]
Thanks for this contribution! Could you rebase this against `main` now that #114 has been merged, and we'll get this reviewed :)

**Review by hcpadkins — CHANGES_REQUESTED** [review]
A few changes, but overall this is a great contribution. Thank you for this one!

File: grove/connectors/sf/threat_detection.py [hunk 1]
```diff
@@ -0,0 +1,709 @@
...
+
+        self.logger.debug(
+            f"Using OAuth token URL: {oauth_token_url} for instance: {self.instance_url}",
+            extra=self.log_context,
+        )
+
+        self.logger.debug(
+            f"OAuth credentials - Client ID: {self.client_id[:8]}... (length: {len(self.client_id) if self.client_id else 0}), "
+            f"Client Secret: {'*' * 8}... (length: {len(self.client_secret) if self.client_secret else 0})",
+            extra=self.log_context,
```
[the remark this answers is not in the document]
Comment (brianbeklaviyo) [reply grove/connectors/sf/threat_detection.py]: [2b6a540](https://github.com/hashicorp-forge/grove/pull/116/commits/2b6a5408713b000566c84cecaf2f98c3093cd9e3)

File: grove/connectors/sf/threat_detection.py [hunk 2]
```diff
@@ -0,0 +1,709 @@
...
+                )
+            except (ValueError, KeyError):
+                # If we can't parse the error response, log the raw response
+                self.logger.error(
+                    f"OAuth authentication failed with unparseable response: {err}",
+                    extra={
+                        **self.log_context,
+                        "oauth_url": oauth_token_url,
+                        "client_id": (
+                            self.client_id[:8] + "..." if self.client_id else None
```
[the remark this answers is not in the document]
Comment (brianbeklaviyo) [reply grove/connectors/sf/threat_detection.py]: [2b6a540](https://github.com/hashicorp-forge/grove/pull/116/commits/2b6a5408713b000566c84cecaf2f98c3093cd9e3)

File: grove/connectors/sf/threat_detection.py [hunk 3]
```diff
@@ -0,0 +1,709 @@
...
+            client = Salesforce(
+                instance_url=instance_url,
+                session_id=session_id,
+                version=SF_VERSION,
+            )
+        except SalesforceError as err:
+            raise RequestFailedException(f"Unable to create Salesforce client: {err}")
+
+        # Check if Salesforce Shield Event Monitoring is available
+        self._check_shield_availability(client)
```
[the remark this answers is not in the document]
Comment (brianbeklaviyo) [reply grove/connectors/sf/threat_detection.py:495]: This was intentional to make the methods easier to test without needing to account for the self.client state in each method test. We can refactor if there is a strong preference to keep the same pattern! However, the current approach makes method dependencies explicit and allows for more isolated unit testing.
